File: ADS PROPOSED PROBLEMS/first problem set.py
def calculate_S1(m):
    result = 0
    for i in range(m+1):
        result += i + (i**2)
    return result

def calculate_S1_recursive(m):
    if m == 0: return 0
    else: return m + m**2 + calculate_S1_recursive(m-1)

def calculate_S2(m, n):
    """nested for loops so pretty much O(n^2)"""
    S2 = 0
    for i in range(1, m + 1):
        for j in range(i, n + 1):
            S2 += j + i
    return S2

def calculate_S3(m, n):
    """nested for loops so pretty much O(n^2)"""
    S3 = 0
    for i in range(1, m + 1):
        for j in range(i, n + 1):
            S3 += (j / i)
    return S3

#/
def f(x): return x**2 
def trapezoidal_rule(func, a, b, n):
    h = (b - a) / n
    result = (func(a) + func(b)) / 2.0

    for i in range(1, n):
        result += func(a + i * h)

    result *= h
    return result

a_value = 0
b_value = 2
n_value = 100

result = round(trapezoidal_rule(f, a_value, b_value, n_value), 4) #rounded to 4 decimals

print(f"The approximate integral using the Trapezoidal Rule is: {result}")
print("_____END_[TRAPEZOIDAL]_____")


# Implementing Simpson's 1/3 
def f(x): return x**2
def simpson13(x0,xn,n):
    h = (xn - x0) / n
    # Finding sum 
    integration = f(x0) + f(xn)
    for i in range(1,n):
        k = x0 + i*h
        if i%2 == 0:
            integration = integration + 2 * f(k)
        else:
            integration = integration + 4 * f(k)
    # Finding final integration value
    integration = integration * h/3
    
    return integration
    
# Input section
lower_limit = 0
upper_limit = 1
sub_interval = 4

# Call trapezoidal() method and get result
result = simpson13(lower_limit, upper_limit, sub_interval)
print("Integration result by Simpson's 1/3 method is: %0.6f" % (result) )

File: ADS PROPOSED PROBLEMS/first problem set.c
#include <stdio.h>

// Problem 1
int calculate_S1(int m) {
    int result = 0;
    for (int i = 0; i <= m; i++) {
        result += i + (i * i);
    }
    return result;
}

int calculate_S1_recursive(int m) {
    if (m == 0)
        return 0;
    else
        return m + m * m + calculate_S1_recursive(m - 1);
}

// Problem 2
int calculate_S2(int m, int n) {
    int S2 = 0;
    for (int i = 1; i <= m; i++) {
        for (int j = i; j <= n; j++) {
            S2 += j + i;
        }
    }
    return S2;
}

// Problem 3
float calculate_S3(int m, int n) {
    float S3 = 0.0;
    for (int i = 1; i <= m; i++) {
        for (int j = i; j <= n; j++) {
            S3 += (float)j / i;
        }
    }
    return S3;
}

// Problem 4 (Trapezoidal Rule)
float f(float x) {
    return x * x;
}

float trapezoidal_rule(float (*func)(float), float a, float b, int n) {
    float h = (b - a) / n;
    float result = (func(a) + func(b)) / 2.0;

    for (int i = 1; i < n; i++) {
        result += func(a + i * h);
    }

    result *= h;
    return result;
}

int main() {
    int m_value = 5;
    int n_value_s2 = 4;
    int n_value_s3 = 4;

    printf("Problem 1: S1(%d) = %d\n", m_value, calculate_S1(m_value));
    printf("Problem 1 (Recursive): S1(%d) = %d\n", m_value, calculate_S1_recursive(m_value));
    printf("Problem 2: S2(%d, %d) = %d\n", m_value, n_value_s2, calculate_S2(m_value, n_value_s2));
    printf("Problem 3: S3(%d, %d) = %f\n", m_value, n_value_s3, calculate_S3(m_value, n_value_s3));

    // Problem 4 (Trapezoidal Rule)
    float a_value = 0.0;
    float b_value = 2.0;
    int n_value_trapezoidal = 100;

    float result_trapezoidal = trapezoidal_rule(f, a_value, b_value, n_value_trapezoidal);
    printf("The approximate integral using the Trapezoidal Rule is: %.4f\n", result_trapezoidal);

    return 0;
}
